### PR TITLE
Fix encoder-decoder ONNX merge

### PR DESCRIPTION
**Pull request by fxmarty**
Fixes https://github.com/huggingface/optimum/issues/921

In next PRs:
1. Add the support for merged decoder for seq2seq models in `optimum-cli export onnx`
2. Add the support for merged decoder for seq2seq models in ORTModel

### CODE REVIEW

File: optimum/onnx/transformations_utils.py [hunk 1]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
     Unifies the outputs of two ONNX model protos. The outputs of model1 will be replaced by outputs of model2.
     According to the rules of "If" op, two subgraphs must have the same number of outputs.
     """
-    _check_num_outputs(model1, model2)
+
+    model1_outputs = {output.name for output in model1.graph.output}
+    model2_outputs = {output.name for output in model2.graph.output}
+
+    if len(model1_outputs) != len(model2_outputs):
+        if strict is True:
```
Comment (michaelbenayoun) [inline optimum/onnx/transformations_utils.py:147]: :)

File: optimum/onnx/transformations_utils.py [hunk 2]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
+                f"The two models proto have different outputs ({len(model1_outputs)} and {len(model2_outputs)} outputs)."
+                " Constant outputs will be added to unify the two models outputs."
+            )
+
+    outputs_only_in_1 = model1_outputs - model2_outputs
+    outputs_only_in_2 = model2_outputs - model1_outputs
+    if len(outputs_only_in_2) > 0:
+        raise ValueError("The ModelProto model2 should not have more outputs than model1.")
+
+    if strict is True and model1_outputs != model2_outputs:
```
Comment (michaelbenayoun) [inline optimum/onnx/transformations_utils.py:163]: :)

File: optimum/onnx/transformations_utils.py [hunk 3]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
-                    f" model protos have same outputs, have same data types and are in the same order."
+                if strict is False and model_output_1.name not in model2_outputs:
+                    data_type = model_output_1.type.tensor_type.elem_type
+                    dims_output_1 = _infer_output_shape(model_output_1)
+                    if not isinstance(dims_output_1[0], str):
+                        raise ValueError(
+                            f"Expected a dynamic shape for the axis zero of {model_output_1.name}, found a static shape: {dims_output_1[0]}"
+                        )
+
+                    # fill the constant shape with the original shape, except for the axis zero that is 0 for an empty constant,
```
Comment (michaelbenayoun) [inline optimum/onnx/transformations_utils.py:184]: :)

File: optimum/onnx/transformations_utils.py [hunk 4]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
+
+                    constant_empty_output = onnx.helper.make_tensor_value_info(
+                        model_output_1.name,
+                        model_output_1.type.tensor_type.elem_type,
+                        _infer_output_shape(model_output_1),
+                    )
+                    model2.graph.output.insert(idx, constant_empty_output)
+                else:
+                    raise ValueError(
+                        f"Can not match {model_output_1.name} with {model_output_2.name}. Make sure your"
```
Comment (michaelbenayoun) [inline optimum/onnx/transformations_utils.py:216]:
```suggestion
                        f"Cannot match {model_output_1.name} with {model_output_2.name}. Make sure your"
```

File: optimum/onnx/transformations_utils.py [hunk 5]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
-        if not model_output_1 == model_output_2:
+        if model_output_1 != model_output_2:
             if not (
                 model_output_1.name == model_output_2.name
                 and model_output_1.type.tensor_type.elem_type == model_output_2.type.tensor_type.elem_type
             ):
-                raise ValueError(
-                    f"Can not match {model_output_1.name} with {model_output_2.name}. Make sure your"
-                    f" model protos have same outputs, have same data types and are in the same order."
+                if strict is False and model_output_1.name not in model2_outputs:
```
Comment (michaelbenayoun) [inline optimum/onnx/transformations_utils.py:176]: :)

File: optimum/onnx/transformations_utils.py [hunk 6]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
+                    model2.graph.output.insert(idx, constant_empty_output)
+                else:
+                    raise ValueError(
+                        f"Can not match {model_output_1.name} with {model_output_2.name}. Make sure your"
+                        f" model protos have same outputs, have same data types and are in the same order."
+                    )
+            else:
+                model2.graph.output.remove(model_output_2)
+
+                # We use model1 (normally the decoder) for the output shape
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/onnx/transformations_utils.py:222]: <3

File: optimum/onnx/transformations_utils.py [hunk 7]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
+        else:
+            logger.info(
+                f"The two models proto have different outputs ({len(model1_outputs)} and {len(model2_outputs)} outputs)."
+                " Constant outputs will be added to unify the two models outputs."
+            )
+
+    outputs_only_in_1 = model1_outputs - model2_outputs
+    outputs_only_in_2 = model2_outputs - model1_outputs
+    if len(outputs_only_in_2) > 0:
+        raise ValueError("The ModelProto model2 should not have more outputs than model1.")
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/onnx/transformations_utils.py:161]: `merge_decoders` is asymmetrical for now - it was before as well as we use the outputs of model1. It's indeed not good and can be changed in an other PR.

File: optimum/onnx/transformations_utils.py [hunk 8]
```diff
@@ -128,44 +134,100 @@ def _infer_output_shape(output: ValueInfoProto):
...
+            )
+        else:
+            logger.info(
+                f"The two models proto have different outputs ({len(model1_outputs)} and {len(model2_outputs)} outputs)."
+                " Constant outputs will be added to unify the two models outputs."
+            )
+
+    outputs_only_in_1 = model1_outputs - model2_outputs
+    outputs_only_in_2 = model2_outputs - model1_outputs
+    if len(outputs_only_in_2) > 0:
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/onnx/transformations_utils.py:160]: Thanks, fixed with set checks!

**Pull request merged**